apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/RetryHandler.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright 2022 Google Inc. All Rights Reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Apis.Storage.v1;
+using Google.Apis.Http;
+using Google.Apis.Json;
+using Google.Apis.Util;
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Google.Cloud.Storage.V1
+{
+    /// <summary>
+    /// Handler for exponential backoff for request messages which permit retries.
+    /// This is similar to BackoffHandler in Google.Apis.Http, but that doesn't give us enough flexibility.
+    /// This handler does not enforce a maximum number of retries; that's configured via <see cref="ConfigurableMessageHandler.NumTries"/>.
+    /// </summary>
+    internal sealed class RetryHandler : IHttpUnsuccessfulResponseHandler
+    {
+        private static readonly string[] s_retriableErrors = { "rateLimitExceeded", "backendError" };
+
+        private static RetryHandler s_instance = new RetryHandler();
+
+        private RetryHandler() { }
+
+        internal static void MarkAsRetriable<TResponse>( StorageBaseServiceRequest<TResponse> request) =>
+            request.AddUnsuccessfulResponseHandler(s_instance);
+
+        internal static async Task<bool> IsRetriableResponse(HttpResponseMessage response)
+        {
+            StandardResponse<object> errorResponse = null;
+            try
+            {
+                // Always retry 502 errors.
+                if (response.StatusCode == HttpStatusCode.BadGateway)
+                {
+                    return true;
+                }
+
+                // For other errors, check the response content.
+                var str = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+                errorResponse = NewtonsoftJsonSerializer.Instance.Deserialize<StandardResponse<object>>(str);
+                var errors = errorResponse.Error?.Errors;
+                // If we couldn't deserialize properly, or the result doesn't contain any reasons, we won't retry.
```

Comment:
If we couldn't deserialize properly, we'll be in the catch block, right?
Basically the comment doesn't seem to match the condition - it looks like this is really "Only retry if we know that all errors specified in the response are retriable"

Reply:
And I note this is a problem in the BigQuery one as well - let's improve it in both places :)